data/simpy.gram

```diff
@@ -44,15 +44,19 @@ break_stmt: a='break' { _Py_Break(EXTRA(a, a)) }
 continue_stmt: a='continue' { _Py_Continue(EXTRA(a, a)) }
 
 import_stmt: import_name | import_from
-import_name: 'import' dotted_as_names
+import_name[stmt_ty]: a='import' b=dotted_as_names { _Py_Import(b, EXTRA(a, b)) }
 # note below: the ('.' | '...') is necessary because '...' is tokenized as ELLIPSIS
 import_from: ('from' (('.' | '...')* !'import' dotted_name | ('.' | '...')+)
               'import' ('*' | '(' import_as_names ')' | import_as_names))
 import_as_name: NAME ['as' NAME]
-dotted_as_name: dotted_name ['as' NAME]
+dotted_as_name[alias_ty]:
+    | a=dotted_name 'as' b=NAME { _Py_alias(((expr_ty) a)->v.Name.id, ((expr_ty) b)->v.Name.id, p->arena) }
+    | a=dotted_name { _Py_alias(((expr_ty) a)->v.Name.id, NULL, p->arena) }
 import_as_names: import_as_name (',' import_as_name)* [',']
-dotted_as_names: dotted_as_name (',' dotted_as_name)*
-dotted_name: NAME ('.' NAME)*
+dotted_as_names[asdl_seq*]: a=dotted_as_name b=further_dotted_as_names* { seq_insert_in_front(p, a, b) }
+further_dotted_as_names[alias_ty]: ',' a=dotted_as_name { a }
+dotted_name[expr_ty]: a=NAME b=further_dotted_name* { seq_to_dotted_name(p, seq_insert_in_front(p, a, b)) }
+further_dotted_name[expr_ty]: '.' a=NAME { a }
```

Comment:
Can you reorder the rules so they appear top-down?

Also `import_as_name[s]` should probably get different names, since they are components of `import_from`, not further variants of import statements. (If you get my drift. If you don't, let me know and I'll explain further.)

Finally, it should be possible to have actions inside parentheses, so you could get rid of `further_*`.

And an idea for later: I wonder if the pattern `X (sep X)*` is going to be so common that we may want to special-case this in the grammar, so we can return this as a sequence of Xs without needing to call `seq_insert_in_front()` (which makes an extra copy).

Reply:
I'll do the reordering, change the name for `import_as_name[s]` and try to remove all the unnecessary `further_*` rules.

On your last idea, quickly going through the grammar I found at least 16 occurrences of this pattern. Do you this it is enough to justify this special case in the grammar? I think, I'm more on the yes side.

Reply:
Great! Please open a separate issue to discuss the design of such a feature.